src/GitHub.Api/Authentication/LoginManager.cs

```diff
@@ -174,25 +174,33 @@ string password
 
             if (ret.IsSuccess)
             {
-                return ret.Output[0];
+                return new LoginResultData(LoginResultCodes.Success, null, host, ret.Output[0]);
             }
 
-            if (ret.IsCustom && ret.Status == "2fa")
+            if (ret.IsTwoFactorRequired)
             {
-                keychain.SetToken(host, ret.Output[0]);
-                await keychain.Save(host);
-                throw new TwoFactorRequiredException();
+                return new LoginResultData(LoginResultCodes.CodeRequired, "Two Factor Required.", host, ret.Output[0]);
+            }
+
+            if (ret.IsBadCredentials)
+            {
+                return new LoginResultData(LoginResultCodes.Failed, "Bad credentials.", host, ret.Output[0]);
+            }
+
+            if (ret.IsLocked)
+            {
+                return new LoginResultData(LoginResultCodes.LockedOut, "Account locked.", host, ret.Output[0]);
             }
 
             if (ret.Output.Any())
             {
-                throw new Exception(string.Join(Environment.NewLine, ret.Output));
+                return new LoginResultData(LoginResultCodes.Failed, "Failed.", host, ret.Output[0]);
             }
 
-            throw new Exception("Authentication failed");
+            return new LoginResultData(LoginResultCodes.Failed, "Failed.", host);
         }
 
-        private async Task<string> TryContinueLogin(
```

Comment:
This method and the `TryLogin` method are exactly the same except for an error message, a flag and the extra line written to the process. Maybe dedupe this too?

Reply:
Consolidated